modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxHandler.java

```diff
@@ -218,73 +222,151 @@ public IgniteTxHandler(GridCacheSharedContext ctx) {
+        ctx.io().addCacheHandler(TX_MSG_HND_ID, GridDhtTxFinishRequest.class, new CI2<UUID, GridCacheMessage>() {
             @Override public void apply(UUID nodeId, GridCacheMessage msg) {
                 processDhtTxFinishRequest(nodeId, (GridDhtTxFinishRequest)msg);
             }
         });
 
-        ctx.io().addCacheHandler(0, GridDhtTxOnePhaseCommitAckRequest.class, new CI2<UUID, GridCacheMessage>() {
+        ctx.io().addCacheHandler(TX_MSG_HND_ID, GridDhtTxOnePhaseCommitAckRequest.class, new CI2<UUID, GridCacheMessage>() {
             @Override public void apply(UUID nodeId, GridCacheMessage msg) {
                 processDhtTxOnePhaseCommitAckRequest(nodeId, (GridDhtTxOnePhaseCommitAckRequest)msg);
             }
         });
 
-        ctx.io().addCacheHandler(0, GridDhtTxFinishResponse.class, new CI2<UUID, GridCacheMessage>() {
+        ctx.io().addCacheHandler(TX_MSG_HND_ID, GridDhtTxFinishResponse.class, new CI2<UUID, GridCacheMessage>() {
             @Override public void apply(UUID nodeId, GridCacheMessage msg) {
                 processDhtTxFinishResponse(nodeId, (GridDhtTxFinishResponse)msg);
             }
         });
 
-        ctx.io().addCacheHandler(0, GridCacheTxRecoveryRequest.class,
+        ctx.io().addCacheHandler(TX_MSG_HND_ID, GridCacheTxRecoveryRequest.class,
             new CI2<UUID, GridCacheTxRecoveryRequest>() {
                 @Override public void apply(UUID nodeId, GridCacheTxRecoveryRequest req) {
                     processCheckPreparedTxRequest(nodeId, req);
                 }
             });
 
-        ctx.io().addCacheHandler(0, GridCacheTxRecoveryResponse.class,
+        ctx.io().addCacheHandler(TX_MSG_HND_ID, GridCacheTxRecoveryResponse.class,
             new CI2<UUID, GridCacheTxRecoveryResponse>() {
                 @Override public void apply(UUID nodeId, GridCacheTxRecoveryResponse res) {
                     processCheckPreparedTxResponse(nodeId, res);
                 }
             });
+
+        ctx.io().addCacheHandler(TX_MSG_HND_ID, ConsistentCutAwareMessage.class,
+            new CI2<UUID, ConsistentCutAwareMessage>() {
+                @Override public void apply(UUID nodeId, ConsistentCutAwareMessage msg) {
+                    processConsistentCutAwareMessage(nodeId, msg);
+                }
+            });
+    }
+
+    /** */
+    private void processConsistentCutAwareMessage(UUID nodeId, ConsistentCutAwareMessage msg) {
+        ctx.snapshotMgr().handleConsistentCutId(msg.cutId());
+
+        setTransactionCutIdIfNeeded(msg);
+
+        GridCacheMessage cacheMsg = msg.payload();
+
+        ctx.io()
+            .cacheHandler(TX_MSG_HND_ID, cacheMsg.getClass())
+            .apply(nodeId, cacheMsg);
+    }
+
+    /**
+     * Set received Consistent Cut ID to transaction if specified.
+     *
+     * @param msg Finish message signed with Consistent Cut ID.
+     */
+    private void setTransactionCutIdIfNeeded(ConsistentCutAwareMessage msg) {
```

Comment:
Please, rename to `setTransactionCutIdIfNeeded` -> `setTransactionCutIdIfRequired`